### PR TITLE
[common-2]: Create User Security Api

### PR DESCRIPTION
**Pull request by Sunagatov**
**1. Task's name:** 
[common-2] Create User Security Api

**2. Developer's full name:** 
Zufar Sunagatov

**3. Link to the PR:** 
https://github.com/Sunagatov/Online-Store/pull/66

**4. Link to Documentation:** 
Created task to create this docs https://trello.com/c/dswPa5Za/98-docs-4-add-usersecurity-api-documentation

**5. Changes:**
- `UserSecurityEndpoint` endpoint with endpoints `register`, `authenticate`, `getUserFromSecurityContext`, `logout` was added.
- `UserAuthenticationService` service with method `authenticate` was added.
- `UserRegistrationService` service with method `register` was added.
- `UserFromSecurityContextProvider` service with methods  `getUserFromSecurityContext` was added.
- `UserSecurityManager` service with methods `register`, `authenticate`, `getUserFromSecurityContext`, `logout` was added.

**6. Tests:**
Tested with Postman by making HTTP GET request to GetShoppingSession endpoint to get the shopping session for the user with the userId = '11111111-1111-1111-1111-111111111111'

**7. Test data:**
HTTP POST request
http://localhost:8083/api/v1/auth/authenticate

{
    "username": "johndoe",
    "password": "pass123"
}

And then using JWT from response:

HTTP GET request
http://localhost:8083/api/v1/products/a3c4d3f7-1172-4fb2-90a9-59b13b35dfc6

**8. Conclusion:**
Project was built with Apache Maven successfully.
Project was tested manually with Postman successfully.

### CODE REVIEW

**Review by UladzislauM — APPROVED** [review]

File: src/main/java/com/zufar/onlinestore/security/configuration/SpringSecurityConfiguration.java [hunk 1]
```diff
@@ -46,6 +54,21 @@ public SecurityFilterChain securityFilterChain(final HttpSecurity httpSecurity,
...
+    @Bean
+    public UserDetailsService userDetailsService(final UserRepository userRepository,
+                                                 final PasswordEncoder passwordEncoder) {
+        return username -> {
+            UserEntity user = userRepository.findUserByUsername(username);
+            if (user == null) {
+                log.warn("Failed to get the user with the username = {}.", username);
+                throw new UsernameNotFoundException(username);
+            }
+            user.setPassword(passwordEncoder.encode(user.getPassword()));
```
Comment (Sanya124) [inline src/main/java/com/zufar/onlinestore/security/configuration/SpringSecurityConfiguration.java:66]: Is it encoding a password and returning a response with a password, but encoded? Why is the password needed here?

**Review by Reyzis2021 — APPROVED** [review]

File: src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java [hunk 2]
```diff
@@ -9,15 +9,15 @@
 @RequiredArgsConstructor
 public class RegistrationDtoConverter {
 
-    public UserDto toDto(final RegistrationRequest registrationRequest) {
+    public UserDto toDto(final UserRegistrationRequest userRegistrationRequest) {
         return new UserDto(
                 null,
```
Comment (shpali4) [inline src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java:14]: may be its need to remove?

File: src/main/java/com/zufar/onlinestore/security/api/UserSecurityManager.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    public UserDto getUserFromSecurityContext() {
+        return userFromSecurityContextProvider.getUser();
+    }
+
+    public void logout(final HttpServletRequest request,
+                       final HttpServletResponse response) {
+        SecurityContextLogoutHandler securityContextLogoutHandler = new SecurityContextLogoutHandler();
+        securityContextLogoutHandler.logout(request, response, null);
+    }
```
Comment (shpali4) [inline src/main/java/com/zufar/onlinestore/security/api/UserSecurityManager.java:38]: do we need here log massage about completed logout?

**Review by Kaadosh — APPROVED** [review]

File: src/main/java/com/zufar/onlinestore/security/endpoint/UserSecurityEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    public static final String USER_SECURITY_API_URL = "/api/v1/auth/";
+
+    private final UserSecurityManager userSecurityManager;
+
+	@PostMapping("/register")
+	public ResponseEntity<UserRegistrationResponse> register(@RequestBody @NotNull @Valid final UserRegistrationRequest request) {
+        UserRegistrationResponse authenticationResponse = userSecurityManager.register(request);
+		return ResponseEntity
+				.ok(authenticationResponse);
```
Comment (vitaliibredun) [inline src/main/java/com/zufar/onlinestore/security/endpoint/UserSecurityEndpoint.java:38]: I suggest to use status 201, because of the new data creation

**Review by vitaliibredun — APPROVED** [review]
Approved

**Review by yevr19 — APPROVED** [review]

**Pull request merged**